Feature/private events

smcommits opened the pull request:
# Description

In this project, I have implemented a site similar to a private Eventbrite which allows users to create events and then manage user signups. Users can create events and send invitations. Other users can only attend events if they have an invitation.

## Model Relations
- User can create many events
- User can attend many events
- Events can be created by a user
- Event can be attended by many users

## Logical Approach
- Login
  To log in, the application checks if the username provided by the user exists in the database and creates a new session id.
- Logout
  The application sets the current_session_id to nil
- Invites
 The application uses enum to create the status column in EventAttendee Database with corresponding strings of "invited" and "accepted". A default EventAttendee instance is created with the default status of invited and changed to accepted when a user accepts the requests.


## How Has This Been Tested?
We have written unit tests and integration tests using Rspec and Capybara
- Validation tests for the model User
-  Association tests for the models' User, Event, and EventAttendee
- Created Integration tests for authentication_management and event_management using Capybara

smcommits committed:
Create User model, controller, and view.

- Created a controller for Users with actions new, show, and create
- Created model for Users
- Created view for new and show action of Users controller
- Created a database migration for Users

DeliceLydia committed:
Add basic event

- Added event model
- Added association to the user model
- Added association to the event model
- added foreign key to the event model
- Added class name to the user model
- Added class name to the event model

smcommits committed:
Implement Login

- Implement the login functionality by creating a sessions_controller.
  Their are three actions defind in sessions_controller; new, create and
  destroy. The create action is using session hash and storing the user
  if he or she is authenticated

Modify User show page

- Modified the user's show page to list all events the user has created

Create Events Controller, corresponding routes; create, show, and index

DeliceLydia committed:
Add event attendance

- Added event_attendee model

in the user model I added:
- Has_many event_attendees association
- Has_many attended_events association

In the event model i added:

- has_many event_attendee association
- has_many attendees association

In the event_attendee model i added:
- belongs_to attendee association
- belongs_to attended_event association

smcommits committed:
Create events show page, User's show page

Event Show Page
- Created events show page to display list of all attendees

User's Show Page

- Created user show page to display all the events has created
- Created logic in view to display all the past attended events, and
  all the upcoming events. Accomplished this using past? and future?
  predicate method of datetime object

- Created a partail for displaying with event title, event date and
  event desciption:

DeliceLydia committed:
Modify index page

- Modified index page to list all events
- Used a class method to separate upcoming and past events
- Refactored past and future events into scopes

smcommits committed:
Add bootstrap, add login functionality, add bootstrap styling

Add Bootstrap
- Added bootstrap to the  project using yarn. Added the bootstrap entry
  to webpacker, application/js/environment, application/stylesheet

Add Login Functionalty
- Add login functionality using sessions_controller and the session has.
  Added methods to check if the user is loggen in by checking the
  session had and added method to validare required_login

Add Bootstrap Styling

- Add bootstrap classes to the elements on index page and user show page

DeliceLydia committed:
Add bootsrap styling on the forms

- Add styling on login form
- Add styling on the signup form
- Add styling on the create event form

smcommits committed:
Add Flash Messages

Added flash messages partial into the layouts application.html.erb.
Created partials to be rendered default with classes corresponding to
the name of bootstrap alert classes.

Added an extension for bootstrap danger and warning class

smcommits committed:
Add Invite Functionality

- Added invite functionality by using enums in EventAttendee model. Used
enums to generate inivted and accepted columns in EventAttendees table.

- Implemented logic to initiate invitation by placing an invite link on the
  events page which user created. The events page takes event_id
  parameter and redirects to all users page

- Implemented logic to send invite to users by creating an invite link
  on all users page and sending event_id and user_id as a param to
  create an instance of event_attendee with a default status of invited

- Implemented logic to accept invitations and display events
  accordingly. Invitation acception is triggered by an accept button on
  users request page. The button updates the corresponding
  event_attendee column to have status of accepted

DeliceLydia committed:
Add tests for associations

- Added tests for user model
- Added tests for event model
- Added tests for event_attendee

smcommits committed:
Write Unit test for User Validtion

- Added unit test for user validation

Write Integration Test for event_management and authentication
management
- Added test check to verify that user is unable to login if workng
  credentials provided
- Added test to verify that user is not able to create an event without
  loggin in
- Added test to verify that user is able to create events after loggin
  in

smcommits committed:
Fix linter errors

smcommits committed:
Fix merger conflict

smcommits committed:
Fix linter errors for stylelint

DeliceLydia committed:
I add screeshot

- I added a screenshot to the README

acushlakoncept commented on db/migrate/20210112125011_create_event_attendees.rb:
```diff
@@ -0,0 +1,9 @@
+class CreateEventAttendees < ActiveRecord::Migration[6.1]
+  def change
+    create_table :event_attendees do |t|
+      t.belongs_to :attendee
+      t.belongs_to :attended_event
+      t.timestamps
+    end
+  end
```
Add an index to your attendee and attended_event, since it allows data to be quickly located without having to search every row in a database table every time a database table is accessed
```
      t.belongs_to :attendee, index: true
      t.belongs_to :attended_event, index: true
```

smcommits replied:
Added Index by creating a new migration "AddIndexToEventsAttendee"

acushlakoncept commented on app/controllers/events_controller.rb:
```diff
@@ -0,0 +1,26 @@
+    @event = Event.find(params[:id])
+  end
+
+  def event_params
+    params.require(:event).permit(:title, :event_date, :description)
+  end
```
Since this method is accessed only in this file, it is usually a good practice to make sure it's private
```
private

def event_params
...

```

smcommits replied:
Moved the event_params method in the private interface of the Event Controllers class.

acushlakoncept commented on app/views/users/show.html.erb:
```diff
@@ -0,0 +1,54 @@
+
+    <h3 class="fw-bold">User Events</h3>
+    <div class="row">
+      <% @user.created_events.each do |event| %>
+        <div class="col-6">
+        <%= render partial: "layouts/event", locals: {event: event} %>
+        </div>
+      <%end%>
+    </div>
+  </div>
+  <hr>
+  <div class="container mt-3">
+  <h3 class="fw-bold">Attending</h3>
+  <div class="row mt-2">
+
+    <div class="col-6">
+
+      <div>
+        <h3>Previously Attended Events</h3>
+        <%= display_past_events(User.find(session[:current_user_id])) %>
+      </div>
+
+    </div>
+
+    <div class="col-6">
+
+      <div>
+        <h3>Upcoming Events</h3>
+        <%@user.attended_events.each do |event| %>
+          <%= display_upcoming_events(@user.id, event) %>
+        <%end%>
+      </div>
```
- [ ] There should be NO application logic in the views nor helpers (no Post.new, Comment.create, current_user.friends.find_by(...)...). Everything connected with data processing should be encapsulated in models and used in controllers.
Inside your User Controller, you can have something like `
@created_events = current_user.events` then on the view you do `@created_events do |event| ...`
You can create a helper method in your application controller, `current_user = User.find(session[:current_user_id])` and use it anywhere in your application.

smcommits replied:
Replaced the data queries from users/show.html.erb and moved them to the show action of Users controller

acushlakoncept commented on app/views/layouts/application.html.erb:
```diff
@@ -11,6 +11,38 @@
+
+       <div class="navbar-nav">
+         <ul class="navbar-nav ml-auto">
+           <%unless logged_in? %>
+             <li class="nav-item"><%= link_to "Login", login_path, class: "nav-link"%></li>
+             <li class="nav-item"><%= link_to "Sign Up", new_user_path, class: "nav-link" %></li>
+           <%end%>
+           <% if logged_in? %>
+             <li class="nav-item"><%= link_to user.username, user_path(session[:current_user_id]), class: "nav-link"%></li>
+             <li class="nav-item"><%= link_to "Log Out", logout_path, method: :delete, class: "nav-link"%></li>
+           <%end%>
```
* [ ] There should be absolutely NO logic in views. >Views are only for displaying information that became available to it from its corresponding controller not generating it. 
View-related logic can be encapsulated in View helpers (app/helpers/*_helper.rb)
**Resources**
* [Where Do I Put My Code?](http://codefol.io/posts/Where-Do-I-Put-My-Code)
* [The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers)
* [Rails View Helpers Best Practices](https://medium.com/@acushla4real/rails-view-helpers-best-practices-guide-d7d7b6db3db4)
* [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright

smcommits replied:
Removed login from views/application.html.erb by adding helper methods in application helpers to create a decision structure based on if the user is logged in or not.

acushlakoncept reviewed (changes requested):
# Status: Changes Required ♻️

### Hi @smcommits and @DeliceLydia, both of you did wonderfully well.

Your readme file looks professional 👏  and commits messages are quite descriptive, keep up the good work 👍 

The application works as expected, congratulations! 🍷 

However, there are still some changes that are necessary for this project to be approved. Address them and request another code review.

* [ ] lists events previoulsy attendented by user **using event model scope**, i.e. `@past_events = current_user.attended_events.previous` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
* [ ] lists upcoming events previously attended by user **using event model scope**, i.e. `@upcoming_events = current_user.attended_events.upcoming` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

Kind check below for other comments I made on your codebase.

### Cheers and Happy coding! 💻 👏

(Feel free to contact me on slack(@acushla), if you have any additional question)
Uduak Essien
[Twitter](https://twitter.com/acushlakoncept) | [LinkedIn](https://www.linkedin.com/in/acushlakoncept/)

smcommits commented:
@acushlakoncept

Thank you so much for the feedback. It helps us to  improve the code. 
I am just confused about the request to use a scope to filter the past or future events attended by the user. I am already using the predefined scope in the Event model to filter the past and future events attended by a user. 


https://github.com/smcommits/private-events/blob/b9312b3da48fb0394494c17a939d1c1c2ffc3fe4/app/helpers/users_helper.rb#L20-L30

Can you please elaborate on the change requested?

smcommits committed:
Implement changes requested

- Added event_params method to private interface
- Added index to event_attendee table
- Replaced data query from users/show.html.erb to show action in user
  controller
- Removed logic from application.html.erb by creating helper methods in
  application_helpers

mahmoud717 reviewed (changes requested):
### Changes required ♻️ 

**Hello @smcommits  & @DeliceLydia**

Great job with the project so far it looks amazing, but there are still some changes to be made:
 
---
### Requested 🔴 

- [ ] make sure to add a sign up link in the login page
- [ ] in the create page. the date field is supposed to be a date field. or how are you going to differentiate past events from future ones. right now if I enter a wrong date the event doesn't show up after creation.
- [ ]  the description field should be a text area.

---

please make these changes and request another code review.

if you have any questions don't hesitate to contact me on slack **MahmoudMohammad**

smcommits commented:
@mahmoud717 Thank you so much for the feedback. 

I am assuming that there is confusion. So, please allow me to elaborate, the _check_invitaion_ and _past_ methods used in user helper are actually scopes that are defined in the event_attendee model. I am not directly querying the model but instead, I am using predefined scopes.

smcommits commented:
@mahmoud717 The login page has already has a signup link
![Screenshot from 2021-01-18 19-20-19](https://user-images.githubusercontent.com/59168430/104924333-479e8d00-59c3-11eb-9d77-2e47329ed6f7.png)

DeliceLydia committed:
Impliment change requested

- Change event_date from text_field to a date_field
- Change description from text field to text_area

jubaan reviewed (changes requested):
# Status: Changes Required :recycle: 

There are some issues that need to be addressed before we can move forward. Please review the list below.

## Required changes :wrench: 
### [Mandatory] :red_circle: 

- [ ] The application lets the user create events without a date. Please add a validator to it so the user can create an event only if the event has at least a title and date.


## Keep up the great work :muscle: and Happy Coding :smiley: :computer:!!

smcommits committed:
Add validation for events

DeliceLydia committed:
Fixed rubocop errors

Georjane reviewed (approved):
## Status: PR Approved 💯 🥇  
### Hi there Team ✋🏽 
Congratulations on successfully completing this project. You have met all requirements for this project to be approved. It's a thumbs up for me. 👍🏽 
You can go ahead and merge this PR.🤝 

Good luck on your next project and Happy Coding 💻 

**Witah Georjane
Code Reviewer at Microverse Inc
[portfolio](https://witahgeorjane.tech/) | [linkedIn](https://www.linkedin.com/in/witah-georjane/)**
